src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -44,65 +36,106 @@ public LinuxInstanceManager(IHttpClientFactory httpClientFactory, IScriptWebHost
+        private async Task<bool> AssignInstanceAsyncCore(HostAssignmentContext context)
         {
-            if (!_webHostEnvironment.InStandbyMode)
+            _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+
+            // set a flag which will cause any incoming http requests to buffer
+            // until specialization is complete
+            // the host is guaranteed not to receive any requests until AFTER assign
+            // has been initiated, so setting this flag here is sufficient to ensure
+            // that any subsequent incoming requests while the assign is in progress
+            // will be delayed until complete
+            _webHostEnvironment.DelayRequests();
+
+            // start the specialization process in the background
+            await AssignAsync(context);
+            return true;
+        }
+
+        public async Task<bool> HandleWarmupRequestAsync(HostAssignmentContext context)
+        {
+            // Based on profiling download code jit-ing holds up cold start.
+            // Pre-jit to avoid paying the cost later.
+            await DownloadWarmupAsync(context.GetRunFromPkgContext());
+            return true;
+        }
+
+        public async Task<bool> AssignInstanceAsync(HostAssignmentContext context)
+        {
+            if (!IsValidEnvironment(context))
             {
-                // This is only true when specializing pinned containers.
-                if (!context.Environment.TryGetValue(EnvironmentSettingNames.ContainerStartContext, out string startContext))
+                return false;
+            }
+
+            if (context.IsWarmupRequest)
+            {
+                return await HandleWarmupRequestAsync(context);
+            }
+
+            lock (_assignmentLock)
+            {
+                if (_assignmentContext != null)
                 {
-                    _logger.LogError("Assign called while host is not in placeholder mode and start context is not present.");
-                    return false;
+                    return _assignmentContext.Equals(context);
                 }
+                _assignmentContext = context;
             }
 
-            if (_environment.IsContainerReady())
+            await AssignInstanceAsyncCore(context);
+
+            return true;
+        }
+
+        public bool StartAssignment(HostAssignmentContext context)
+        {
+            if (!IsValidEnvironment(context))
             {
-                _logger.LogError("Assign called while container is marked as specialized.");
                 return false;
             }
 
             if (context.IsWarmupRequest)
             {
-                // Based on profiling download code jit-ing holds up cold start.
-                // Pre-jit to avoid paying the cost later.
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
+                _ = HandleWarmupRequestAsync(context);
```

Comment:
Do we know if this is guaranteed to be async? If not, may be safer to leave it in a `Task.Run`.